Content.Server/Nyanotrasen/StationEvents/Events/GlimmerRevenantSpawnRule.cs

```diff
@@ -1,32 +1,71 @@
+using System.Linq;
 using Robust.Shared.Random;
+using Robust.Shared.Map;
 using Content.Server.GameTicking.Rules.Components;
 using Content.Server.Psionics.Glimmer;
 using Content.Server.StationEvents.Components;
+using Content.Shared.Psionics.Glimmer;
 
 namespace Content.Server.StationEvents.Events;
 
-internal sealed class GlimmerRevenantRule : StationEventSystem<GlimmerRevenantRuleComponent>
+internal sealed class GlimmerRevenantSpawnRule : StationEventSystem<GlimmerRevenantRuleComponent>
 {
-    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly IRobustRandom _robustRandom = default!;
+    [Dependency] private readonly GlimmerSystem _glimmerSystem = default!;
+
+    private static readonly string RevenantPrototype = "MobRevenant";
+    private static readonly string WispPrototype = "MobGlimmerWisp";
+    private static readonly string ShadePrototype = "MobShade";
 
     protected override void Started(EntityUid uid, GlimmerRevenantRuleComponent component, GameRuleComponent gameRule, GameRuleStartedEvent args)
     {
         base.Started(uid, component, gameRule, args);
 
-        List<EntityUid> glimmerSources = new();
+        var glimmerSources = EntityManager.EntityQuery<GlimmerSourceComponent, TransformComponent>().ToList();
+        var normalSpawnLocations = EntityManager.EntityQuery<VentCritterSpawnLocationComponent, TransformComponent>().ToList().ConvertAll(item => item.Item2.Coordinates);
+        var hiddenSpawnLocations = EntityManager.EntityQuery<MidRoundAntagSpawnLocationComponent, TransformComponent>().ToList().ConvertAll(item => item.Item2.Coordinates);
+
+        var spawnLocations = normalSpawnLocations.ToHashSet();
+        spawnLocations.UnionWith(hiddenSpawnLocations.ToHashSet());
+
+        int total = 1;
+
+        if (_glimmerSystem.GetGlimmerTier() == GlimmerTier.Critical)
+            total = 2;
 
-        var query = EntityQueryEnumerator<GlimmerSourceComponent>();
-        while (query.MoveNext(out var source, out _))
+        int i = 0;
+        while (i < total)
         {
-            glimmerSources.Add(source);
-        }
+            if (glimmerSources.Count != 0 && _robustRandom.Prob(0.9f))
+            {
+                SpawnRevenant(_robustRandom.Pick(glimmerSources).Item2.Coordinates);
+                i++;
+                continue;
+            }
+
+            if (normalSpawnLocations.Count != 0)
+            {
+                SpawnRevenant(_robustRandom.Pick(spawnLocations));
+                i++;
+                continue;
+            }
```

Comment:
I understand total never exceeds 2, but this looks odd.

If it were a normal loop and not a construct to avoid code duplication, I'd say any logic regarding the length of a list of candidates should be done before entering a loop. This is checking the length of enumerable data every loop without altering it.

You can at least put the increment at the boundary of the loop, change to an if/else-if statement and remove both continue statements, and it'd make it look less odd.

Finally, you can also get rid of `i` and just use `total` as a decrementer.